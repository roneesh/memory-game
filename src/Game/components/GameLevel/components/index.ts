import LevelReady from './LevelReady';
// import LevelPlaying from './LevelPlaying';
// import LevelFinished from './LevelFinished';
// import LevelPreview from './LevelPreview';

export {
  LevelReady,
  // LevelPlaying,
  // LevelFinished,
  // LevelPreview,
};
