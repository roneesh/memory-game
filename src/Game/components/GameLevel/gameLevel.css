.game-level {
  margin-bottom: 20px;
}

.game-board {
  background-color: #4f5d75;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  padding: 3px 0;
}

.game-board__tile {
  width: 100px;
  height: 100px;
  margin: 3px;
  perspective: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #bfc0c0;
  color: #ffffff;
  font-size: 32px;
  font-family: 'Monoton', cursive;
}

@keyframes tileWiggle {
  0%   { transform:  rotateZ(0deg) }
  25%  { transform:  rotateZ(5deg) }
  50%  { transform:  rotateZ(-5deg) }
  100% { transform:  rotateZ(0deg) }
}

.game-board__tile--wiggling {
  position: relative;
  animation-name: tileWiggle;
  animation-duration: 3s;
}

@keyframes tileExplodeUp {
  0%   { top: 0px }
  15%  { top: 100px }
  40%  { top: 1000px }
  100% { top: 10000px }
}

@keyframes tileExplodeDown {
  0%   { top: 0px }
  15%  { top: -100px }
  40%  { top: -1000px }
  100% { top: -10000px }
}

.game-board__tile--exploding-up {
  position: relative;
  animation-name: tileExplodeUp;
  animation-duration: 3s;
}

.game-board__tile--exploding-down {
  position: relative;
  animation-name: tileExplodeDown;
  animation-duration: 3s;
}
