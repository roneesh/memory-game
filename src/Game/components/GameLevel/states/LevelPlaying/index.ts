import LevelPlaying from './LevelPlaying';

export default LevelPlaying;
