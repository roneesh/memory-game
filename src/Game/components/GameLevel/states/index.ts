// import LevelFinished from './LevelFinished';
// import LevelPlaying from './LevelPlaying';
import LevelPreview from './LevelPreview';
import LevelReady from './LevelReady';

export {
  // LevelFinished,
  // LevelPlaying,
  LevelPreview,
  LevelReady,
};
