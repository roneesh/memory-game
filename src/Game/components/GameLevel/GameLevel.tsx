import { shuffle } from 'lodash';
import * as React from 'react';
import './gameLevel.css';
import { LevelFinished, LevelPlaying, LevelPreview, LevelReady, } from './states';

interface IGameLevelProps {
  updateLevel: () => void;
  level: number;
};

type LevelState = "ready" | "preview" | "playing" | "finished";
export interface IBoardTile {
  character: string | HTMLElement;
};
export interface ILevelTurns {
  count: number;
  currentTurn: number[];
}
interface IGameLevelState {
  levelState: LevelState;
  board: IBoardTile[];
  turns:  ILevelTurns;
};

class GameLevel extends React.Component<IGameLevelProps, IGameLevelState> {
  constructor(props: IGameLevelProps) {
    super(props);

    this.state = {
      board: this.generateBoard(),
      levelState: 'ready',
      turns: {
        count: 0,
        currentTurn: [],
      },
    };
  }

  public generateBoard(): IBoardTile[] {
    return shuffle(
      ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
      'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'].map(character => {
        return {
          character,
        };
      })
    );
  }

  public setLevelState = (newLevelState: LevelState) => {
    if (newLevelState === 'ready') {
      this.setState({
        board: this.generateBoard(),
        levelState: newLevelState,
        turns: {
          count: 0,
          currentTurn: [],
        },
      });
    }
    this.setState({ levelState: newLevelState });
  }

  public getResetBtn() {
    const { levelState } = this.state;

    if (levelState === 'finished') {
      return null;
    }

    const text = levelState === 'ready' ? 'Play Level' : 'Reset';
    const nextState = levelState === 'ready' ? 'preview' : 'ready';
    return <button onClick={this.setLevelState.bind(null, nextState)}>
      {text}
    </button>;
  }

  public renderControls() {
    return <div className="game-level__controls">
      {this.getResetBtn()}
    </div>
  }

  public renderState() {
    const { levelState, board } = this.state;
    const { updateLevel } = this.props;

    switch(levelState) {
      case 'ready':
        return <LevelReady board={board} />;
      case 'preview':
        return <LevelPreview board={board} setLevelState={this.setLevelState} />;
      case 'playing':
        return <LevelPlaying board={board} setLevelState={this.setLevelState} />;
      case 'finished':
        return <LevelFinished board={board} updateLevel={updateLevel} />;
      default:
        return null;
    }
  }

  public componentDidUpdate(prevProps: IGameLevelProps) {
    if (this.props.level !== prevProps.level) {
      this.setLevelState('ready');
    }
  }
  public render() {
    return <div className="game-level">
      {this.renderControls()}
      {this.renderState()}
    </div>;
  }
}

export default GameLevel;
