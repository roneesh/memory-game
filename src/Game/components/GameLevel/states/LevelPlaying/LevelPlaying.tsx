import * as React from 'react';
import { IBoardTile } from '../../GameLevel';


interface ILevelPlayingProps {
  board: IBoardTile[];
  setLevelState: (newLevelState: string) => void;
};

interface ILevelPlayingTile {
  character: string | HTMLElement;
  status: 'unfound' | 'found' | 'selected';
}
interface ILevelTurns {
  count: number;
  selections: number[];
};
interface ILevelPlayingState {
  boardCopy: ILevelPlayingTile[];
  turns: ILevelTurns;
}

class LevelPlaying extends React.Component<ILevelPlayingProps, ILevelPlayingState> {
  constructor(props: ILevelPlayingProps) {
    super(props);

    this.state = {
      boardCopy: Array.from(props.board).map(tile => {
        return {
          character: tile.character,
          status: 'unfound',
        } as ILevelPlayingTile;
      }),
      turns: {
        count: 0,
        selections: [],
      },
    };
  }

  public onTileClick = async (index: number) => {
    const { boardCopy, turns } = this.state;
    const tile = boardCopy[index];

    const maxTilesSelected = turns.selections.length === 2;
    const tileIsAlreadyClicked = tile.status === 'selected' || tile.status === 'found';
    const dontProcessClick = maxTilesSelected || tileIsAlreadyClicked;
    if (dontProcessClick) {
      return;
    }

    tile.status = 'selected';
    const newSelections = [...turns.selections, index];
    const newTurns = {
      count: this.state.turns.count + 1,
      selections: newSelections,
    };

    if (newSelections.length < 2) {
      this.setState({ turns: newTurns });
    } else {
      this.setState({ turns: newTurns }, this.resolveTurn);
    }
  }

  public resolveTurn = async () => {
    // 200ms delay before we hide tiles again
    await new Promise(resolve => setTimeout(() => {resolve()}, 200));

    const { boardCopy, turns } = this.state;

    const selections = turns.selections;
    const tileCharactersMatch =
      boardCopy[selections[0]].character === boardCopy[selections[1]].character;
    if (tileCharactersMatch) {
      boardCopy[selections[0]].status = 'found';
      boardCopy[selections[1]].status = 'found';
    } else {
      boardCopy[selections[0]].status = 'unfound';
      boardCopy[selections[1]].status = 'unfound';
    }
    const allFoundTiles = boardCopy.filter(tile => tile.status === 'found');
    if (allFoundTiles.length === 24) {
      this.props.setLevelState('finished');
    } else {
      this.setState({
        turns: {
          count: turns.count,
          selections: [],
        },
      });
    }
  }

  public render() {
    const { boardCopy, turns } = this.state;

    return <div>
      <div className="game-board">
        {boardCopy.map((tile, index) => {
          return (
            <div className="game-board__tile"
              key={index}
              onClick={this.onTileClick.bind(this, index)}
            >
              {tile.status === 'unfound' ? null : tile.character}
            </div>
          );
        })}
      </div>
      <div>
        <p>Turns: {turns.count} </p>
      </div>
    </div>
  }

};

export default LevelPlaying;
