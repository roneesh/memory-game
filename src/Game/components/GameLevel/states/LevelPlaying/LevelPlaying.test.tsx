import { shallow } from 'enzyme';
import * as React from 'react';
import renderer from 'react-test-renderer';
import { IBoardTile } from '../../GameLevel';
import LevelPlaying from './LevelPlaying';

describe('The level when its playing', () => {
  let board: IBoardTile[] = [];
  let setLevelStateSpy: any;
  const boardBase =
    ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
    'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'];

  const getComponent = () => shallow((<LevelPlaying board={board} setLevelState={setLevelStateSpy} />));
  const getSnapshot = () => <LevelPlaying board={board} setLevelState={jest.fn()} />;

  beforeEach(() => {
    setLevelStateSpy = jest.fn();
    board = boardBase.map(character => {
      return {
        character,
      };
    });
  });

  it('renders a blank board on load', () => {
    const levelPlaying = renderer
      .create(getSnapshot())
      .toJSON();
    expect(levelPlaying).toMatchSnapshot();
  });

  it('on click of a blank tile it counts it as a turn and makes it a visible state', () => {
    const component = getComponent();
    const instance = component.instance() as LevelPlaying;
    component.find('.game-board__tile').first().simulate('click');

    expect(instance.state.turns.count).toEqual(1);
  });

  it('doesnt count an already clicked tile as a second turn or new tile', () => {
    const component = getComponent();
    const instance = component.instance() as LevelPlaying;
    const firstTile = component.find('.game-board__tile').first();
    firstTile.simulate('click');
    firstTile.simulate('click');

    expect(instance.state.turns.count).toEqual(1);
  });

  // it('dispatches the next statit(e update when its foundPairs lenght is 12', () => {

  // });
});
