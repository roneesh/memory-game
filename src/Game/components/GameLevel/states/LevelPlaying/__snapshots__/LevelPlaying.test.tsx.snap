// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The level when its playing renders a blank board on load 1`] = `
<div>
  <div
    className="game-board"
  >
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
    <div
      className="game-board__tile"
      onClick={[Function]}
    />
  </div>
  <div>
    <p>
      Turns: 
      0
       
    </p>
  </div>
</div>
`;
